### ISSUE TITLE
Support multiple lookup types using one index

### ISSUE THREAD

**Issue by fabiankessler** [fmmfonseca/completely#4]
There are different concepts built-in already for index lookups. It's very flexible. Some examples:

* *exact match, lower case*:
  using a HashMultiMap in the Index, EqualityAutomaton in the index lookup, and a LowerCaseTransformer as the Analyzer in the engine
* *starts-with exact match, lower case*: 
  using a PatriciaTrie in the Index, EqualityAutomaton in the index lookup, and a LowerCaseTransformer as the Analyzer in the engine
* *stats-with fuzzy match, lower case*: 
  using a PatriciaTrie in the Index, EditDistanceAutomaton in the index lookup, and a LowerCaseTransformer as the Analyzer in the engine

In my use case, with a million of indexed entries, I want to perform "exact starts-with" matching first. If that brings good results, fine, I take it. If not, then I go ahead and thry the "fuzzy starts-with" matching.

For this I currently either need 2 indexes (not an option, and technically not necessary), or some ugly syntax.

The problem is that the IndexLookupStrategy.lookup() method is not flexible. One option would be to pass in a closure how I want the lookup to be done. Then it could be controlled from the outside.

My current solution is this:
Create a PatriciaTrie instance.
Create 2 instances of IndexAdapter, give both the same trie. One uses the EqualityAutomaton and the other the EditDistanceAutomaton.
Create 2 engines, one per IndexAdapter, 
Now feed my data for indexing only to 1 engine.
Now both engines are ready for querying.

What do you think?

**Comment by fmmfonseca** [comment]
Fabian,

I don't think creating two indexes is necessary for your use case, specially since you intend to use the same Analyzer, they will be exactly the same.

Your logic in on the query side, I see two options:

1. Create an index adapter with the logic you describe:

```java
class SampleAdapter implements IndexAdapter<SampleRecord>
{
    private FuzzyIndex<SampleRecord> index = new PatriciaTrie<SampleRecord>();

    @Override
    public Collection<SampleRecord> get(String token)
    {
        Collection<SampleRecord> result = index.getAll(token);
        if (result.isEmpty())
        {
            result = index.getAny(new EditDistanceAutomaton(token, Math.log(token.length())));
        }
        return result;
    }

    @Override
    public boolean put(String token, SampleRecord value)
    {
        return index.put(token, value);
    }
}
```

This assumes, every token will be searched individually for exact and approximate matches...

2. If what you want is to perform a regular search and only after a second approximate search, together with custom sorting, you might me better off creating a wrapper class over two AutocompleteEngine's

**Comment by fabiankessler** [comment]
Option 1 does not work for me. Because as you mentioned I need custom sorting.

But you're right, maybe a wrapper around the engine would work best. I might try that later.

For me, it would probably also work to perform the sorting within the lookup() method. And then perform another search wiht sorting if the first was not good enough. But that's only possibly because in my use case I will always only deal with 1 token (no splitting).

**Comment by fmmfonseca** [comment]
The third option is to write a custom Engine altogether, it's only a facade for indexing and searching after all...

**Comment by fmmfonseca** [comment]
I took a second look at your record sorter. You might be able to avoid a second search if if you perform a single approximate search, and have a RecordWeighter that favours exact query matches, over approximate. The exact will be always be displayed first.

**Comment by fabiankessler** [comment]
Yes, I tried it, but for my case it was better with separate searches. I'll explain:

For short strings (like 3 chars) the approximate search finds thousands of options. My sorting is expensive (it sums up with so many). I'd need a 2 step sorting phase, to get rid of all the fuzzy matches first. And that makes no sense for my use case. It complicates it all. 

I got good performance and results with the 2 searches: 
Most of the times the first exact starts-with search is enough. It finds relatively few, sorting is fast, done. This is what most users want anyway.
If nothing "good" was found nothing at all, or only rare words) then it's likely that the user typed something wrong. The first search found almost nothing. Therefore it finished fast. Now I go with the fuzzy search.
